src/dataflow/src/source/postgres.rs

```diff
@@ -209,11 +209,38 @@ impl PostgresSourceReader {
         let stream = LogicalReplicationStream::new(copy_stream);
         tokio::pin!(stream);
 
+        let mut last_keepalive = Instant::now();
         let mut inserts = vec![];
         let mut deletes = vec![];
         while let Some(item) = stream.next().await {
+            use ReplicationMessage::*;
+            // The upstream will periodically request keepalive responses by setting the reply field
+            // to 1. However, we cannot rely on these messages arriving on time. For example, when
+            // the upstream is sending a big transaction its keepalive messages are queued and can
+            // be delayed arbitrarily.  Therefore, we also make sure to send a proactive keepalive
+            // every 30 seconds.
+            //
+            // See: https://www.postgresql.org/message-id/CAMsr+YE2dSfHVr7iEv1GSPZihitWX-PMkD9QALEGcTYa+sdsgg@mail.gmail.com
+            if matches!(item, Ok(PrimaryKeepAlive(ref k)) if k.reply() == 1)
+                || last_keepalive.elapsed() > Duration::from_secs(30)
```

Comment:
A 10-second interval may be safer here if the timeout that will apply is 60 seconds.

Reply:
Can you explain the failure scenario you have in mind? From what I understand the current code will cause a timeout only if the network has a 30 second latency, in which case a reconnect is probably a better course of action